### PR TITLE
Fix OpLoad pointee type validation.

### PR DESCRIPTION
**Pull request by dekimir**

### CODE REVIEW

**Comment by dekimir** [comment]
@umar456 PTAL

**Comment by umar456** [comment]
LGTM

File: source/validate_id.cpp [hunk 1]
```diff
@@ -779,14 +779,21 @@ bool idUsage::isValid<SpvOpLoad>(const spv_instruction_t* inst,
...
-  auto type = usedefs_.FindDef(pointer.second.words[1]);
-  assert(type.first);
-  spvCheck(resultType.second.id != type.second.id,
-           DIAG(resultTypeIndex)
-               << "OpLoad Result Type <id> '" << inst->words[resultTypeIndex]
-               << " does not match Pointer <id> '" << pointer.second.id
-               << "'s type.";
-           return false);
+  auto pointerType = usedefs_.FindDef(pointer.second.words[1]);
+  if (!pointerType.first || pointerType.second.opcode != SpvOpTypePointer) {
```
Comment (dneto0) [inline source/validate_id.cpp]: As an aside, I'm finding the use of a std::pair return value from usedefs_.FindDef very hard to read.  I'd rather have a bespoke struct with nice field values.

Comment (dneto0) [inline source/validate_id.cpp]: This check for existence of the def and whether it's a pointer type suggests  a short utility function. 
The pattern is also used within OpVariable, OpCopyMemorySized.

Maybe we can do that later, along with a use of a DiagnosticStream object to shorten the return sequences.  But let's wait so we avoid making challenging merges for collaborators.

File: test/ValidateID.cpp [hunk 2]
```diff
@@ -709,7 +709,7 @@ TEST_F(ValidateID, OpLoadResultTypeBad) {
 %5 = OpVariable %3 UniformConstant
 %6 = OpFunction %1 None %4
 %7 = OpLabel
-%8 = OpLoad %2 %5
+%8 = OpLoad %3 %5
```
Comment (dneto0) [inline test/ValidateID.cpp]: Hm.  Should stay as %2  ?!

**Comment by dneto0** [comment]
-1 I think one of the test changes is in correct.  Which begs the question of whether the code is correct.
Also, you're adding diagnostics.  No better time than the present to add tests for those diagnostics.

File: test/ValidateID.cpp (hunk 2 above)
Comment (dekimir) [inline test/ValidateID.cpp]: Why?  That would make it identical to the preceding test.

File: source/validate_id.cpp (hunk 1 above)
Comment (dekimir) [inline source/validate_id.cpp]: I'm OK doing the proposed refactorings in a later PR.

File: test/ValidateID.cpp (hunk 2 above)
Comment (dneto0) [inline test/ValidateID.cpp]: Whoops.  I didn't realize it was supposed to be a *BAD* result.
And the GitHub differ makes it very difficult to see beyond the end of this segment.

**Comment by dneto0** [comment]
+2

**Comment by dneto0** [comment]
Ok.  Your test changes are correct.
I can wait for the diagnostics tests.
+2

**Comment by dekimir** [comment]
Merged via 58852aa012462a0a6da8391374943fd89bfc4151.

**Pull request closed without merging**